### ISSUE TITLE
Pseudo whole genome alignment

### ISSUE THREAD

**Issue by AdrianAllen1977** [TGenNorth/NASP#21]
Hi, have sued NASP in the past and found it very useful. Is ti possible to get a pseudo whole genome alignment from the pipeline?  I want to be able to use one to check from recombination using Gubbins and Clonal Frame.  From memory I think in the past I only used an  output fast from NASP that indicated variant sites. Is there a way of getting variant and invariant sites printed to the final output fasta?

Thank you

Adrian

**Comment by jmonroynieto** [comment]
Hello, Adrian. We haven't received any summons from the court yet ;)

There are three ways to achieve what you want. The explanations here is superficial, so I recommend you check [our publication](https://www.microbiologyresearch.org/content/journal/mgen/10.1099/mgen.0.000074) if my writing is confusing. In particular, the fourth paragraph of the methods explains each matrix.

### NASP Solutions

1) You may be able to produce this alignment from external genomes, which will be aligned to the reference via NUCmer. You will find these files with the filename extension `.frankenfasta`.  For an explanation on frankenfasta see the bottom of this reply.

2) If you'd like to include a cleaner form that can also includes read files, you could produce the master matrix (master-masked is recommended), which includes all calls based on the reference genome.  To generate the master matrix, include `--withallrefpos` flag if you are accessing the nasptool directly, or answer `y` to the last question in the dispatcher questionnaire. 

As a reminder, any matrix can be transformed into FASTA or VCF formats using `nasp export`

### Other Solutions  

3) If you are looking to only produce genome alignments, you could use (mummer)[https://github.com/mummer4/mummer] and use `delta2maf` to produce an alignment file.

### Frankenfasta usage
```
usage: frankenfasta [mummer deltas]

Frankenfasta creates a fasta 1-to-1 position aligned with a reference fasta
by combining a sample fasta with a nucmer alignment delta file.

How it works:

To start with, a list of reference fasta contigs and their lengths are read from
the delta file. This is used to create a fasta where all positions are masked
with 'X'.

	>Contig Description
	XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
	XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
	XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
	XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
	XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
	XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
	XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
	XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
	XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
	XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
	XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
	XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX

The delta alignments are used to copy positions from the sample fasta into their
corresponding positions in the reference fasta.

	>Contig Description
	XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
	XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
	XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXAAAAAAAXAAXXAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
	AAAAAGTTACTTTCATTAATAGAGCAAAATTTATTAATTATACTTTTTACACAAAAAACAAAGAGAGGATCGACCTTTTC
	TATTCATTCTTTAAACTACAGATTCTTTATAAAAAAATTTAACGACGTTGAGAGTAGACATGAGCGAAAGCACGGGTAGT
	AGCACGCTTCAACTTGGATTGAGCGGCCTTGAGGTCATCAGCTTGAGCATCGTCAGCGGTGAGTTCCATCAATTCCATGG
	CTTCAGCGAGAGCAGTTTCAATGGCTTCACGGTCACCACGACGGAGCTTCATGTTGACGTTAGGGTCAGAGATGTTGTTC
	TCAACTTGGTAGACGTAAGATTCAAGATCTTGCTTGGCTTGTACAACGGCTTCACGTTGCTTGTCAGATTCAGCATTCTT
	TTCAGCATCTTGAACCATACGTTCGATATCAGCAGCAGAGAGACGAGTAGAGTTGGAGATGGTGACATCAGCCTTGCGAC
	CAGTGGTCTTGTCTTGAGCAGTGACCTTCAAGATACCATTGGCATCCAAGTCAAAGGTACAGAGAAGTTCAGGGGTACCA
	CGGGGAGCAGGGATAAGACCAGAGAGAGAGAACTCACCGAGAAGGTTGTTTTCCTTGGTCATCAAACGTTCACCTTCGTA
	GACAGGGAAAGTGACAGTGGTTXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX

'X' indicates the absence of calls. Either no sequence aligned to the position,
or it was a deletion.

'N' indicates uncertain calls. Two alignments with different calls may have mapped
to the same position.
```

**Comment by AdrianAllen1977** [comment]
Thanks Juan,
 Ha! I clicked send in my original email and then noticed my misspelling of
'used'. It gave me pause that I'd just sent a mail saying I was 'suing' the
writers of a bioinformatics tool someone in the US had made. Given your
nation's record on all matters related to litigation  I did have a slight
worry....so I'm glad you're being light hearted about it all ;-)

Thank you for your advice and help. I'll reread the paper and also have a
play with the frakenfasta outputs etc.

Cheers again for coming back to me on this. I had the pleasure of working
with Kevin Drees at UNH a few years back and used NASP a lot. Hoping to so
again.

Thanks and very best regards

Adrian
